trio/_threads.py

```diff
@@ -422,6 +427,13 @@ def from_thread_run_sync(fn, *args, trio_token=None):
           "foreign" thread, spawned using some other framework, and still want
           to enter Trio.
     """
+
+    if not callable(fn) or inspect.iscoroutinefunction(fn):
+        raise TypeError(
+            "Trio expected a sync function, but {!r} appears to not be "
+            "callable or asynchronous".format(getattr(fn, "__qualname__", fn))
```

Comment:
You know which it is (is it non-callable or is it an async function?), so the error message should say which it is.

Also, be extra careful about grammatical ambiguity in error messages, since we want them to be as comprehensible as possible. In this case "not be callable or asynchronous" could be read as "not be (callable or asynchronous)" where what you mean is "(not be callable) or (be asynchronous)". I would write that as "is not callable or is an async function".

Reply:
Python already gives a decent error message if you try to call something that's non-callable, so we could drop that check entirely and just let Python raise the exception for us.

Reply:
Oh, whoops, I thought I had left a comment about that but it seems I didn't.

I was also going to suggest that instead of `inspect.iscoroutinefunction` (which won't catch async methods, async functions used with `partial`, etc), you wait until you call the function below and then raise an error if the return value satisfies `inspect.iscoroutine`.